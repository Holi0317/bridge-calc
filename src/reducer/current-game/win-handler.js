// @flow
import {GameStage} from '../../game-stage'
import {computeScores} from './compute-scores'
import {toEndedState} from './converter'
import {bidWinGenerator} from './bid-win-generator'
import {toFront} from '../../utils'

import type {GameState, WaitingBidState, WaitingWinState} from './types'
import type {WIN_ACTION} from '../../actions/current-game'

/**
 * Handler for WIN action
 */
export function winHandler(state: WaitingBidState | WaitingWinState, action: WIN_ACTION): GameState {
  if (state.stage === GameStage.waitingBid) {
    return state
  }
  const win = action.win || state.win
  if (state.rounds === state.currentRound) {
    // Last round
    const newState: WaitingWinState = {
      ...state,
      scores: computeScores(state.bid, win, state.scores)
    }
    return toEndedState(newState, action.time)
  } else {
    const newState: WaitingBidState = {
      ...state,
      stage: GameStage.waitingBid,
      bid: bidWinGenerator(Object.keys(state.names)),
      currentRound: state.currentRound + 1,
      currentPlayerOrder: toFront(state.currentPlayerOrder, 1),
      scores: computeScores(state.bid, win, state.scores)
    }
    delete (newState: any).win
    return newState
  }
}
