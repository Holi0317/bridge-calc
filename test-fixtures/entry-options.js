import { deepFreeze } from "./deep-freeze";

export const defaultState = deepFreeze({
  rounds: 13,
  startingRound: 1,
  playerNames: ["John", "Mary", "Henry", "Joe"],
  importOpened: false
});

/**
 * Create mock redux state tree for test related to entry module.
 *
 * @param {object} rest Object to override default state in return tree
 * @returns {object} partial redux tree for testing
 */
export function makeEntryTree(rest) {
  return {
    entry: {
      ...defaultState,
      ...rest
    }
  };
}
