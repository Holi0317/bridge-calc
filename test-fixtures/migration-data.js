export const round1 = {
  'Bridge.currentRound': '1',
  'Bridge.maker': '0',
  'Bridge.players': '[{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Jonathan"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Helen"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Humphrey"},{"bid":null,"win":null,"_buffer":{"lastRound":null,"combo":null,"miss":null,"maxCombo":null,"sum":null},"score":[],"name":"Margery"}]',
  'Bridge.state': '1',
  'Bridge.totalRounds': '13'
}

export const round2 = {
  'Bridge.currentRound':'2',
  'Bridge.maker':'1',
  'Bridge.players':'[{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":1,"miss":0,"maxCombo":0,"sum":11},"score":[11],"name":"Jonathan"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Helen"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Humphrey"},{"bid":null,"win":null,"_buffer":{"lastRound":1,"combo":0,"miss":1,"maxCombo":0,"sum":-1},"score":[-1],"name":"Margery"}]',
  'Bridge.state':'1',
  'Bridge.totalRounds':'13'
}
