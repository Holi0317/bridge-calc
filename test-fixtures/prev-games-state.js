export function makePrevGamesTree(prevGames) {
  return {
    prevGames: {
      prevGames
    }
  };
}
