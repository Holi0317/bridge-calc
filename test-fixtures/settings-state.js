import {deepFreeze} from './deep-freeze'

export const defaultState = deepFreeze({
  maker: null,
  names: {}
})
