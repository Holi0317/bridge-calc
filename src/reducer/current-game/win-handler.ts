import {GameStage} from '../../game-stage'
import {computeScores} from './compute-scores'
import {toEndedState, toWaitingBidState} from './converter'
import {bidWinGenerator} from './bid-win-generator'
import {toFront} from '../../utils'
import {GameState, IWaitingBidState, IWaitingWinState} from './types'
import {IWinAction} from '../../actions/current-game'

/**
 * Handler for WIN action
 */
export function winHandler(state: IWaitingBidState | IWaitingWinState, action: IWinAction): GameState {
  if (state.stage === GameStage.waitingBid) {
    return state
  }
  const win = action.win || state.win
  if (state.rounds === state.currentRound) {
    // Last round
    const newState = toEndedState(state, action.time)
    return {
      ...newState,
      scores: computeScores(state.bid, win, state.scores)
    }
  } else {
    // Proceed to next round
    const newState = toWaitingBidState(state)
    newState.bid = bidWinGenerator(Object.keys(state.names))
    newState.currentRound = state.currentRound + 1
    newState.currentPlayerOrder = toFront(state.currentPlayerOrder, 1)
    newState.scores = computeScores(state.bid, win, state.scores)
    return newState
  }
}
