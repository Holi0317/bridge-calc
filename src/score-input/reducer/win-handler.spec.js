import {winHandler} from './win-handler'
import {genMap, waitingBidState, waitingWinState} from '../../../test-fixtures/current-game-states'
import {WIN} from '../score-input-actions'

// Just be lazy
const time = new Date(1)

test('it should do no-op when state is waitingBid', () => {
  const state = {
    ...waitingBidState
  }
  const action = {
    type: WIN,
    win: genMap(0, 0, 0, 1),
    time
  }
  const expected = {
    ...waitingBidState
  }
  const actual = winHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should use given win to calculate score', () => {
  const state = {
    ...waitingWinState
  }
  const action = {
    type: WIN,
    win: genMap(0, 0, 0, 1),
    time
  }
  const actual = winHandler(state, action)
  expect(actual).toMatchSnapshot()
})

test('it should use win in state object to calculate score when win is not given in action', () => {
  const state = {
    ...waitingWinState,
    win: genMap(1, 0, 0, 0)
  }
  const action = {
    type: WIN,
    time
  }
  const actual = winHandler(state, action)
  expect(actual).toMatchSnapshot()
})

test('it should end game when it is dispatched at last round', () => {
  const state = {
    ...waitingWinState,
    rounds: 1
  }
  const action = {
    type: WIN,
    win: genMap(1, 0, 0, 0),
    time
  }
  const actual = winHandler(state, action)
  expect(actual).toMatchSnapshot()
})
