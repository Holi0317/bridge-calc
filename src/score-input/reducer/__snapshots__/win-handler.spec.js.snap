// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should end game when it is dispatched at last round 1`] = `
Object {
  "endTime": 1970-01-01T00:00:00.001Z,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 1,
  "scores": Object {
    "a": Array [
      -1,
    ],
    "b": Array [
      10,
    ],
    "c": Array [
      10,
    ],
    "d": Array [
      10,
    ],
  },
  "stage": "ended",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;

exports[`it should use given win to calculate score 1`] = `
Object {
  "bid": Object {
    "a": 0,
    "b": 0,
    "c": 0,
    "d": 0,
  },
  "currentPlayerOrder": Array [
    "b",
    "c",
    "d",
    "a",
  ],
  "currentRound": 2,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "a": Array [
      10,
    ],
    "b": Array [
      10,
    ],
    "c": Array [
      10,
    ],
    "d": Array [
      -1,
    ],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;

exports[`it should use win in state object to calculate score when win is not given in action 1`] = `
Object {
  "bid": Object {
    "a": 0,
    "b": 0,
    "c": 0,
    "d": 0,
  },
  "currentPlayerOrder": Array [
    "b",
    "c",
    "d",
    "a",
  ],
  "currentRound": 2,
  "names": Object {
    "a": "John",
    "b": "Mary",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 13,
  "scores": Object {
    "a": Array [
      -1,
    ],
    "b": Array [
      10,
    ],
    "c": Array [
      10,
    ],
    "d": Array [
      10,
    ],
  },
  "stage": "waitingBid",
  "startTime": 1970-01-01T00:00:00.000Z,
}
`;
